### PR TITLE
WIP: Table constants

### PR DESCRIPTION
**Pull request by hlgp**

### CODE REVIEW

**Comment by hlgp** [comment]
Still a lot to adjust, but could use some feedback on the general idea of where this is headed.

File: warehouse/core/src/main/java/datawave/tables/schema/MetadataFamilyConstants.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package datawave.tables.schema;
+
+public class MetadataFamilyConstants {
+    // a little conflicted about where this should live since metadata-utils has a ColumnFamilyConstants, but I guess having it in two places is better than 40
```
Comment (apmoriarty) [inline warehouse/core/src/main/java/datawave/tables/schema/MetadataFamilyConstants.java:4]: I don't think it's appropriate for this project to depend so heavily on another project's constants. A small change in metadata-utils could cause havoc in here.

File: warehouse/core/src/main/java/datawave/tables/schema/ShardFamilyConstants.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    // locality group name
+    /* TODO Make a clearer definition of full content indexers */
+    public static final String TERM_FREQUENCY_LOCALITY_NAME = "termfrequency";
+    
+    /**
+     * Fi column
+     */
+    public static final String FI = "fi";
+    public static final Text FI_TXT = new Text(FI);
+    public static final byte[] FI_BYTES = FI_TXT.getBytes();
```
Comment (apmoriarty) [inline warehouse/core/src/main/java/datawave/tables/schema/ShardFamilyConstants.java:31]: `FI_PREFIX` is a common constant with a value of `"fi\0"` or `"fi\u0000"`.

This begs the questions, are these constants purely for the column families or are these a collection of table constants?

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/IteratorBuildingVisitorTest.java [hunk 3]
```diff
@@ -169,10 +170,10 @@ public void NeTest() throws Exception {
         Key hit = new Key("row", "dataType" + Constants.NULL + "123.345.456");
         
         List<Map.Entry<Key,Value>> source = new ArrayList<>();
-        source.add(new AbstractMap.SimpleEntry(
-                        new Key("row", "fi" + Constants.NULL + "F1", "v0" + Constants.NULL + "dataType" + Constants.NULL + "123.345.456"), new Value()));
-        source.add(new AbstractMap.SimpleEntry(
-                        new Key("row", "fi" + Constants.NULL + "F1", "v1" + Constants.NULL + "dataType" + Constants.NULL + "123.345.456"), new Value()));
+        source.add(new AbstractMap.SimpleEntry(new Key("row", ShardFamilyConstants.FI + Constants.NULL + "F1", "v0" + Constants.NULL + "dataType"
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/jexl/visitors/IteratorBuildingVisitorTest.java:174]: If we're going to make broad changes, I would have a preference for static imports. 

Quick plug for `FI_PREFIX` and static imports
```java
new Key("row", FI_PREFIX + "F1", "v0" + NULL + "dataType"+ NULL + "123.345.456");
```

**Review by apmoriarty — COMMENTED** [review]
Overall I'm in favor of extracting constants and logically grouping them by table. In practice this won't be as clean as the ColumnFamilyConstants in the metadata-utils that inspired this work. For example, the shard table FI constants are only a portion of the "fi\0FIELD" cf. 

My preference would be to break this into smaller merge requests, but if the consensus is to go with this merge request I'm happy to accommodate.

File: warehouse/core/src/main/java/datawave/tables/schema/ShardFamilyConstants.java (hunk 2 above)
Comment (hlgp) [reply warehouse/core/src/main/java/datawave/tables/schema/ShardFamilyConstants.java:31]: I dig the PREFIX.  Will add.
I was going to start with family related constants and see what crawls out of the woodwork for another MR for other things.

File: warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/error/ErrorDataTypeHandler.java [hunk 4]
```diff
@@ -117,9 +118,9 @@
     // Initialize a default (hash-based) UID builder
     private UIDBuilder<UID> uidBuilder = UID.builder();
     
-    public static final Text EVENT_COLF = new Text("e");
+    public static final Text EVENT_COLF = new Text(MetadataFamilyConstants.E);
     public static final Text INFO_COLF = new Text("info");
```
Comment (ivakegg) [inline warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/error/ErrorDataTypeHandler.java:122]: Why no constant for this one?

**Comment by ivakegg** [comment]
I actually love this PR.  Thank you for consolidating these constants.

File: warehouse/core/src/main/java/datawave/tables/schema/MetadataFamilyConstants.java (hunk 1 above)
Comment (ivakegg) [reply warehouse/core/src/main/java/datawave/tables/schema/MetadataFamilyConstants.java:4]: Should this class live in metadata-utils?

Comment (hlgp) [reply warehouse/core/src/main/java/datawave/tables/schema/MetadataFamilyConstants.java:4]: I am not a fan of moving this alone to metadata-utils.  I think that all table structure / constants related classes should exist in a core converged package, be that in datawave core or a new repo of its own.  I would hope that this work would mean removing ColumnFamilyConstants from metadata-utils in favor of the centralized schema definitions.

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/IteratorBuildingVisitorTest.java (hunk 3 above)
Comment (hlgp) [reply warehouse/query-core/src/test/java/datawave/query/jexl/visitors/IteratorBuildingVisitorTest.java:174]: plug noted

File: warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/error/ErrorDataTypeHandler.java (hunk 4 above)
Comment (hlgp) [reply warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/error/ErrorDataTypeHandler.java:122]: It's a WIP.
Also decided to start with the biggest offenders first.  This one is seldom used.

**Comment by hlgp** [comment]
> My preference would be to break this into smaller merge requests, but if the consensus is to go with this merge request I'm happy to accommodate.

Considering it's largely a find/replace, I think breaking it up is rather silly. 

>  In practice this won't be as clean as the ColumnFamilyConstants in the metadata-utils that inspired this work.

It was never meant to be as clean as CFC, which didn't actually inspire this work.  We define these constants again and again willy-nilly across the board and I'm just trying to get to one place.  This is a baby step towards that.

File: warehouse/query-core/src/test/java/datawave/query/IfThisTestFailsThenHitTermsAreBroken.java [hunk 5]
```diff
@@ -519,14 +520,14 @@ public static void writeItAll(Connector con, WhatKindaRange range) throws Except
                 bw = con.createBatchWriter(TableName.SHARD, bwConfig);
                 mutation = new Mutation(shard);
                 
-                mutation.put("fi\u0000" + "UUID", lcNoDiacriticsType.normalize("First") + "\u0000" + datatype + "\u0000" + firstUID, columnVisibility,
-                                timeStamp, emptyValue);
+                mutation.put(ShardFamilyConstants.FI + "\u0000" + "UUID", lcNoDiacriticsType.normalize("First") + "\u0000" + datatype + "\u0000" + firstUID,
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/IfThisTestFailsThenHitTermsAreBroken.java:523]: `FI_PREFIX` opportunity

File: warehouse/query-core/src/test/java/datawave/query/UseOccurrenceToCountInJexlContextTest.java [hunk 6]
```diff
@@ -453,41 +454,41 @@ public static void writeItAll(Connector con, WhatKindaRange range) throws Except
...
-                mutation.put("fi\u0000" + "BAR", lcNoDiacriticsType.normalize("BAR") + "\u0000" + datatype + "\u0000" + secondUID, columnVisibility, timeStamp,
-                                emptyValue);
-                
-                mutation.put("fi\u0000" + "NAME", lcNoDiacriticsType.normalize("NAME8") + "\u0000" + datatype + "\u0000" + thirdUID, columnVisibility,
-                                timeStamp, emptyValue);
-                mutation.put("fi\u0000" + "NAME", lcNoDiacriticsType.normalize("NAME9") + "\u0000" + datatype + "\u0000" + thirdUID, columnVisibility,
-                                timeStamp, emptyValue);
-                mutation.put("fi\u0000" + "BAR", lcNoDiacriticsType.normalize("BAZ") + "\u0000" + datatype + "\u0000" + thirdUID, columnVisibility, timeStamp,
-                                emptyValue);
+                mutation.put(ShardFamilyConstants.FI + "\u0000" + "UUID", lcNoDiacriticsType.normalize("First") + "\u0000" + datatype + "\u0000" + firstUID,
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/UseOccurrenceToCountInJexlContextTest.java:457]: `FI_PREFIX` opportunity

File: warehouse/query-core/src/test/java/datawave/query/ancestor/AncestorChildExpansionIteratorTest.java [hunk 7]
```diff
@@ -89,7 +90,7 @@ public void testNoTopNextError() throws IOException {
     // end basic iterator contract verification
     
     private static final String FI_ROW = "shard";
-    private static final String FI_COLUMN_FAMILY = "fi" + Constants.NULL_BYTE_STRING + "field";
+    private static final String FI_COLUMN_FAMILY = ShardFamilyConstants.FI + Constants.NULL_BYTE_STRING + "field";
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/ancestor/AncestorChildExpansionIteratorTest.java:93]: `FI_PREFIX` opportunity

File: warehouse/query-core/src/test/java/datawave/query/data/parsers/DatawaveKeyTest.java [hunk 8]
```diff
@@ -65,7 +66,7 @@ public void testShardKeyWithExtraNulls() {
     
     @Test
     public void testFiKeyWithExtraNulls() {
-        Key key = new Key(row, "fi\u0000fieldName", "fieldValue\u0000blah\u0000datatype\u0000uid", cv);
+        Key key = new Key(row, ShardFamilyConstants.FI + "\u0000fieldName", "fieldValue\u0000blah\u0000datatype\u0000uid", cv);
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/data/parsers/DatawaveKeyTest.java:69]: `FI_PREFIX` opportunity. I should also come back and clean up this test..

File: warehouse/query-core/src/test/java/datawave/query/iterator/QueryIteratorIT.java [hunk 9]
```diff
@@ -1235,8 +1236,8 @@ protected Key getFI(String field, String value, String uid, long timestamp) {
     }
     
     protected Key getFI(String row, String field, String value, String dataType, String uid, long timestamp) {
-        return new Key(row, "fi" + Constants.NULL_BYTE_STRING + field.toUpperCase(), value + Constants.NULL_BYTE_STRING + dataType + Constants.NULL_BYTE_STRING
-                        + uid, timestamp);
+        return new Key(row, ShardFamilyConstants.FI + Constants.NULL_BYTE_STRING + field.toUpperCase(), value + Constants.NULL_BYTE_STRING + dataType
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/iterator/QueryIteratorIT.java:1249]: `FI_PREFIX` opportunity

File: warehouse/query-core/src/test/java/datawave/query/iterator/logic/TermFrequencyIndexIteratorTest.java [hunk 10]
```diff
@@ -447,11 +448,11 @@ private SortedListKeyValueIterator getIterator() {
     }
     
     private Key getFiKey(String row, String dataType, String uid, String fieldName, String fieldValue) {
-        return new Key(row, "fi" + Constants.NULL + fieldName, fieldValue + Constants.NULL + dataType + Constants.NULL + uid);
+        return new Key(row, ShardFamilyConstants.FI + Constants.NULL + fieldName, fieldValue + Constants.NULL + dataType + Constants.NULL + uid);
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/iterator/logic/TermFrequencyIndexIteratorTest.java:457]: `FI_PREFIX` opportunity

File: warehouse/query-core/src/test/java/datawave/query/jexl/functions/IdentityAggregatorTest.java [hunk 11]
```diff
@@ -100,7 +101,7 @@ public void apply_testSeek() throws IOException {
     }
     
     private Key getFi(String row, String field, String value, String dataType, String uid, long timestamp) {
-        return new Key(row, "fi" + Constants.NULL_BYTE_STRING + field, value + Constants.NULL_BYTE_STRING + dataType + Constants.NULL_BYTE_STRING + uid,
-                        timestamp);
+        return new Key(row, ShardFamilyConstants.FI + Constants.NULL_BYTE_STRING + field, value + Constants.NULL_BYTE_STRING + dataType
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/jexl/functions/IdentityAggregatorTest.java:104]: `FI_PREFIX` opportunity

File: warehouse/query-core/src/test/java/datawave/query/predicate/TLDEventDataFilterTest.java [hunk 12]
```diff
@@ -106,7 +107,7 @@ public void getCurrentField_fiTest() {
         replayAll();
         
         // expected key structure
-        Key key = new Key("row", "fi" + Constants.NULL + "field", "value" + Constants.NULL + "dataType" + Constants.NULL + "123.345.456");
+        Key key = new Key("row", ShardFamilyConstants.FI + Constants.NULL + "field", "value" + Constants.NULL + "dataType" + Constants.NULL + "123.345.456");
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/predicate/TLDEventDataFilterTest.java]: `FI_PREFIX` opportunity

File: warehouse/query-core/src/test/java/datawave/query/testframework/GroupingAccumuloWriter.java [hunk 13]
```diff
@@ -42,7 +43,7 @@ class GroupingAccumuloWriter {
     private static final Logger log = Logger.getLogger(GroupingAccumuloWriter.class);
     
     private static final String NULL_SEP = "\u0000";
-    private static final String FIELD_INDEX = "fi" + NULL_SEP;
+    private static final String FIELD_INDEX = ShardFamilyConstants.FI + NULL_SEP;
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/testframework/GroupingAccumuloWriter.java:46]: `FI_PREFIX` opportunity

File: warehouse/query-core/src/test/java/datawave/query/tld/TLDFieldIndexAggregatorTest.java [hunk 14]
```diff
@@ -174,7 +175,7 @@ public void apply_testSeek() throws IOException {
     }
     
     private Key getFi(String row, String field, String value, String dataType, String uid, long timestamp) {
-        return new Key(row, "fi" + Constants.NULL_BYTE_STRING + field, value + Constants.NULL_BYTE_STRING + dataType + Constants.NULL_BYTE_STRING + uid,
-                        timestamp);
+        return new Key(row, ShardFamilyConstants.FI + Constants.NULL_BYTE_STRING + field, value + Constants.NULL_BYTE_STRING + dataType
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/tld/TLDFieldIndexAggregatorTest.java:178]: `FI_PREFIX` opportunity

File: warehouse/query-core/src/test/java/datawave/query/tld/TLDTest.java [hunk 15]
```diff
@@ -39,14 +40,14 @@ public class TLDTest {
     
     // Build a Forward-Index Key for the Shard table
     private Key buildFiKey(String uid) {
-        String cf = "fi" + '\u0000' + field;
+        String cf = ShardFamilyConstants.FI + '\u0000' + field;
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/tld/TLDTest.java:43]: `FI_PREFIX` opportunity

File: warehouse/query-core/src/test/java/datawave/query/util/CommonalityTokenTestDataIngest.java [hunk 16]
```diff
@@ -279,79 +280,79 @@ public static void writeItAll(Connector con, WhatKindaRange range) throws Except
...
-                            emptyValue);
-            mutation.put("fi\u0000" + "CAT", lcNoDiacriticsType.normalize("cougar") + "\u0000" + datatype + "\u0000" + myUID, columnVisibility, timeStamp,
-                            emptyValue);
-            mutation.put("fi\u0000" + "CAT", lcNoDiacriticsType.normalize("tiger") + "\u0000" + datatype + "\u0000" + myUID, columnVisibility, timeStamp,
-                            emptyValue);
-            mutation.put("fi\u0000" + "CAT", lcNoDiacriticsType.normalize("leopard") + "\u0000" + datatype + "\u0000" + myUID, columnVisibility, timeStamp,
-                            emptyValue);
-            mutation.put("fi\u0000" + "CAT", lcNoDiacriticsType.normalize("puma") + "\u0000" + datatype + "\u0000" + myUID, columnVisibility, timeStamp,
-                            emptyValue);
+            mutation.put(ShardFamilyConstants.FI + "\u0000" + "CAT", lcNoDiacriticsType.normalize("tabby") + "\u0000" + datatype + "\u0000" + myUID,
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/util/CommonalityTokenTestDataIngest.java:278]: `FI_PREFIX` opportunity

File: warehouse/query-core/src/test/java/datawave/query/util/CompositeTestingIngest.java [hunk 17]
```diff
@@ -255,58 +256,70 @@ public static void writeItAll(Connector con, WhatKindaRange range) throws Except
             mutation = new Mutation(shard);
             // corleones
             // uuid
-            mutation.put("fi\u0000" + "UUID", lcNoDiacriticsType.normalize("One") + "\u0000" + datatype + "\u0000" + oneUUID, columnVisibility, timeStamp,
-                            emptyValue);
+            mutation.put(ShardFamilyConstants.FI + "\u0000" + "UUID", lcNoDiacriticsType.normalize("One") + "\u0000" + datatype + "\u0000" + oneUUID,
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/util/CompositeTestingIngest.java:254]: `FI_PREFIX` opportunity

File: warehouse/query-core/src/test/java/datawave/query/util/LimitFieldsTestingIngest.java [hunk 18]
```diff
@@ -158,35 +159,35 @@ public static void writeItAll(Connector con, WhatKindaRange range) throws Except
             
             mutation = new Mutation(shard);
             
-            mutation.put("fi\u0000" + "FOO_3", lcNoDiacriticsType.normalize("abcd") + "\u0000" + datatype + "\u0000" + myUID, columnVisibility, timeStamp,
-                            emptyValue);
+            mutation.put(ShardFamilyConstants.FI + "\u0000" + "FOO_3", lcNoDiacriticsType.normalize("abcd") + "\u0000" + datatype + "\u0000" + myUID,
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/util/LimitFieldsTestingIngest.java:157]: `FI_PREFIX` opportunity

File: warehouse/query-core/src/test/java/datawave/query/util/VisibilityWiseGuysIngest.java [hunk 19]
```diff
@@ -263,89 +264,90 @@ public static void writeItAll(Connector con, WhatKindaRange range) throws Except
             mutation = new Mutation(shard);
             // corleones
             // uuid
-            mutation.put("fi\u0000" + "UUID", lcNoDiacriticsType.normalize("CORLEONE") + "\u0000" + datatype + "\u0000" + corleoneUID, columnVisibilityEnglish,
-                            timeStamp, emptyValue);
+            mutation.put(ShardFamilyConstants.FI + "\u0000" + "UUID", lcNoDiacriticsType.normalize("CORLEONE") + "\u0000" + datatype + "\u0000" + corleoneUID,
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/util/VisibilityWiseGuysIngest.java:262]: `FI_PREFIX` opportunity

File: warehouse/query-core/src/test/java/datawave/query/util/VisibilityWiseGuysIngestWithModel.java [hunk 20]
```diff
@@ -539,94 +540,98 @@ public static void writeItAll(Connector con, WhatKindaRange range) throws Except
             mutation = new Mutation(shard);
             // corleones
             // uuid
-            mutation.put("fi\u0000" + "UUID", lcNoDiacriticsType.normalize("CORLEONE") + "\u0000" + datatype + "\u0000" + corleoneUID, columnVisibilityItalian,
-                            timeStamp, emptyValue);
+            mutation.put(ShardFamilyConstants.FI + "\u0000" + "UUID", lcNoDiacriticsType.normalize("CORLEONE") + "\u0000" + datatype + "\u0000" + corleoneUID,
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/util/VisibilityWiseGuysIngestWithModel.java:538]: `FI_PREFIX` opportunity

File: warehouse/query-core/src/test/java/datawave/query/util/WiseGuysIngest.java [hunk 21]
```diff
@@ -600,106 +601,124 @@ public static void writeItAll(Connector con, WhatKindaRange range) throws Except
             mutation = new Mutation(shard);
             // corleones
             // uuid
-            mutation.put("fi\u0000" + "UUID", lcNoDiacriticsType.normalize("CORLEONE") + "\u0000" + datatype + "\u0000" + corleoneUID, columnVisibility,
-                            timeStamp, emptyValue);
+            mutation.put(ShardFamilyConstants.FI + "\u0000" + "UUID", lcNoDiacriticsType.normalize("CORLEONE") + "\u0000" + datatype + "\u0000" + corleoneUID,
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/util/WiseGuysIngest.java:599]: `FI_PREFIX` opportunity

**Review by apmoriarty — COMMENTED** [review]
This is looking good overall. Understanding that there are outstanding changes, I dropped some notes for where `FI` could become `FI_PREFIX` if you wanted to do some additional work -- feel free to ignore and resolve those comments should you decline that work.

**Review by ivakegg — APPROVED** [review]

**Comment by hlgp** [comment]
I will do this work in a new MR using the datawave-tables project

**Pull request closed without merging**